src/HIE/Bios/Cradle.hs

```diff
@@ -1212,6 +1227,43 @@ readProcessWithOutputs outputNames l workDir cp = flip runContT return $ do
             removeFileIfExists file
             action (name, file)
 
+      reportProgress :: CompilationProgressReporter -> String -> CabalParserState -> IO (CabalParserState, String)
+      reportProgress Nothing str cps = pure (cps, str)
+      reportProgress (Just reporter) str cps@(CabalParserToBuild { numPackagesDeclared = numPackages }) = do
+        let startBuilding = do reporter (CompilationProgress { numPackagesToCompile = numPackages
+                                                             , numPackagesCompiled = 0
+                                                             })
+                               pure (CabalParserBuilding { numPackagesCompleted = 0, numPackagesToBuild = numPackages }, str)
+        case str of
+           ' ':'-':' ':_ -> pure (cps { numPackagesDeclared = numPackages + 1 }, str)
+           'S':'t':'a':'r':'t':'i':'n':'g':' ':' ':' ':' ':' ':_ -> startBuilding
+           _ -> pure (cps, str)
+      reportProgress (Just reporter) str cps@(CabalParserBuilding { numPackagesCompleted = numPackages
+                                                                  , numPackagesToBuild = totalPackages
+                                                                  }) =
+        case str of
+           'C':'o':'m':'p':'l':'e':'t':'e':'d':' ':' ':' ':' ':_ -> do
+             reporter (CompilationProgress { numPackagesToCompile = totalPackages
+                                           , numPackagesCompiled = numPackages + 1
+                                           })
+             pure (cps { numPackagesCompleted = numPackages + 1 }, str)
+           _ -> pure (cps, str)
```

Comment:
I think running a process shouldn't be coupled this tightly with a particular build tool. How about we pull this out and pass the progress reporter to the function?

Reply:
Good point. It is much more elegant if I just pass an accumulator function, it also highlighted that I was passing the string through the function unnecessarily. Fixed here: https://github.com/haskell/hie-bios/pull/436/commits/6d71b932eab6b38c6f07d6b4a648cdea7e866fe2